File: Sep/Day11/sort-vowels-in-a-string.py
#  Time: O(N * LogN)  ::  Space: O(N)
class Solution:
    #TC, SC - O(n log n), O(n)
    def isVowel(self, c: str) -> bool:
        return c in "aeiouAEIOU"

    def sortVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU")
        arr = list(s)

        # Collect all vowels
        key = [ch for ch in arr if self.isVowel(ch)]

        # Sort vowels
        key.sort()

        # Replace vowels in order
        j = 0
        for i in range(len(arr)):
            if self.isVowel(arr[i]):
                arr[i] = key[j]
                j += 1

        return ''.join(arr)



#  Time: O(N)  :: Space: O(1)
class Solution:
    def isVowel(self, c: str) -> bool:
        return c in "aeiouAEIOU"

    def sortVowels(self, s: str) -> str:
        from collections import Counter

        # Count frequencies of vowels
        count = Counter([c for c in s if self.isVowel(c)])

        # Sorted vowel order (ASCII)
        sortedVowel = "AEIOUaeiou"

        ans = []
        j = 0
        for ch in s:
            if not self.isVowel(ch):
                ans.append(ch)
            else:
                # Move to the next vowel that still has remaining count
                while j < len(sortedVowel) and count[sortedVowel[j]] == 0:
                    j += 1
                ans.append(sortedVowel[j])
                count[sortedVowel[j]] -= 1

        return "".join(ans)

File: Sep/Day11/sort-vowels-in-a-string.cc
//  Time: O(N * LogN)  ::  Space: O(N)
class Solution {
public:
    bool isVowel(char c) {
        string vowels = "aeiouAEIOU";
        return vowels.find(c) != string::npos;
    }
    string sortVowels(string s) {
        vector<char> key;
        for (char c : s) {
            if (isVowel(c)) key.push_back(c);
        }

        sort(key.begin(), key.end());

        int j = 0;
        for (int i = 0; i < s.size(); i++) {
            if (isVowel(s[i])) {
                s[i] = key[j++];
            }
        }
        return s;
    }
};



//  Time: O(N)  :: Space: O(1)
class Solution2 {
public:
    bool isVowel(char c) {
        string vowels = "aeiouAEIOU";
        return vowels.find(c) != string::npos;
    }
    string sortVowels(string s) {
        unordered_map<char, int> count;
        for (char c : s) {
            if (isVowel(c)) count[c]++;
        }

        string sortedVowel = "AEIOUaeiou";
        string ans = "";
        int j = 0;

        for (char c : s) {
            if (!isVowel(c)) {
                ans.push_back(c);
            } else {
                while (j < sortedVowel.size() && count[sortedVowel[j]] == 0) {
                    j++;
                }
                ans.push_back(sortedVowel[j]);
                count[sortedVowel[j]]--;
            }
        }
        return ans;
    }
};
